### PR TITLE
Fix issues multiselect

### PR DESCRIPTION
**Pull request by Trakkasure**
This resolves issues noted in:
#44
#43
#33
#30

### CODE REVIEW

**Comment by Trakkasure** [comment]
I signed it!

**Comment by danopia** [comment]
I just upgraded 0.5 => 1.0 and struggled for hours trying to databind `selectedValues` correctly. My singular lists were working, but nothing would make the multi-lists update when the user selected something.

This patchset fixes my problems so it would be great if this (or a similar PR) got merged so I don't have to store a polymer patch in my repo

**Comment by Tanbouz** [comment]
Fixes include iron-selectable for issue #30 ?

**Comment by Trakkasure** [comment]
@Tanbouz Indeed after looking closer, I did find the issue that was causing this one. I added one more for #30 after fully testing the example.

**Comment by Trakkasure** [comment]
All of these fixes are available on my working_production branch until such time as they are merged into the main code in this main repo.

File: iron-multi-selectable.html [hunk 1]
```diff
@@ -61,11 +61,9 @@
         if (this.selectedValues) {
           this._toggleSelected(value);
         } else {
-          this.selectedValues = [value];
+          this.set('selectedValues',[value]);
```
Comment (cdata) [inline iron-multi-selectable.html]: Why is this change necessary?

File: iron-multi-selectable.html [hunk 2]
```diff
@@ -61,11 +61,9 @@
...
         }
-      } else {
-        this.selected = value;
```
Comment (cdata) [inline iron-multi-selectable.html]: Ditto, re: https://github.com/PolymerElements/iron-selector/pull/48/files#r38454974

File: iron-multi-selectable.html [hunk 3]
```diff
@@ -89,6 +87,19 @@
       }
     },
 
+    _observeItems: function(node) {
+      var observer = new MutationObserver(function() {
+        if ((this.multi && this.selectedValues && this.selectedValues.length) || this.selected != null)  {
```
Comment (cdata) [inline iron-multi-selectable.html]: Instead of completely overriding the implementation of this method _just_ to change the boolean logic here, we should probably factor out the boolean to an accessor. This way, overriding implementation can replace the accessor without having to re-implement the `MutationObserver` part of this method. Such an approach would have the bonus effect of making this conditional much easier to read.

File: iron-multi-selectable.html [hunk 4]
```diff
@@ -103,9 +114,11 @@
       var i = this.selectedValues.indexOf(value);
       var unselected = i < 0;
       if (unselected) {
-        this.selectedValues.push(value);
+        this.push('selectedValues',value);
```
Comment (cdata) [inline iron-multi-selectable.html]: :+1:

File: iron-multi-selectable.html [hunk 5]
```diff
@@ -103,9 +114,11 @@
...
+        //this.set('selected',value);
```
Comment (cdata) [inline iron-multi-selectable.html]: Please remove this comment.

File: iron-multi-selectable.html [hunk 6]
```diff
@@ -103,9 +114,11 @@
...
       } else {
-        this.selectedValues.splice(i, 1);
+        this.splice('selectedValues',i,1);
+        //this.set('selected',this.selectedValues[this.selectedValues.length-1]||null);
```
Comment (cdata) [inline iron-multi-selectable.html]: Please remove this comment.

File: iron-selection.html [hunk 7]
```diff
@@ -32,7 +32,7 @@
      * the selected item or undefined if there is no selection.
      */
     get: function() {
-      return this.multi ? this.selection : this.selection[0];
+      return this.multi ? this.selection.slice() : this.selection[0];
```
Comment (cdata) [inline iron-selection.html]: Is this still necessary with the updates made to https://github.com/PolymerElements/iron-selector/pull/48/files#diff-5c8329bac0158c065b259270a43cfe7bR117 ?

**Comment by cdata** [comment]
Thanks for the contribution! It is looking pretty good. I noticed that you are marking several issues as resolved, but the change does not include any testing that might catch regressions in the future. Would you please consider adding such tests?

File: iron-multi-selectable.html (hunk 1 above)
Comment (Trakkasure) [inline iron-multi-selectable.html]: Once I made the change to explicitly call the setter, the notifications were being called. I believe firefox was misbehaving.

File: iron-multi-selectable.html (hunk 2 above)
Comment (Trakkasure) [inline iron-multi-selectable.html]: same misbehaving as above

File: iron-selection.html (hunk 7 above)
Comment (Trakkasure) [inline iron-selection.html]: Yes. I believe returning a copy will guarantee data integrity. If the reference is returned, data changed on that object, then someone uses the element accessors to change data in the component, then unexpected behavior results. This way, the element's selection integrity is maintained.

**Comment by Trakkasure** [comment]
Absolutely. I'll add tests.

File: iron-multi-selectable.html (hunk 1 above)
Comment (cdata) [inline iron-multi-selectable.html]: If there is a browser inconsistency in the way databinding behaves, this should be reported as an issue against Polymer core. I will create a test case to verify what you are saying.

**Comment by cdata** [comment]
@Trakkasure FYI I am going to fork this and add relevant tests. Thanks again for the contribution.

**Comment by cdata** [comment]
Merged these changes with #68 .
